woocommerce/woocommerce-ios
# Shipping Labels: state machine, address validation in shipping labels form, suggested address logics

Part of #2973

## Description
In this PR, many things have been handled under the hood, so the changes visible to the user are few.
- [x] the validation address endpoint will be called also in the main Shipping Label Form, for validating the origin/destination address when the user presses **Continue** button.
- [x] If the address is validated, the next step in the shipping label form will be enabled (implemented the state machine).
- [x] If there is a suggested address, the new screen that will allow the users to choose the suggested address will be shown (WIP, will be completed in the next PR).
- [x] If the user edits an existing address, after the validation the next step in the shipping label form will be enabled.

## Changes
- Implemented a new `UIButton` subclass called `ButtonActivityIndicator` that is able to display or hide an activity indicator in the middle.
- The new button `ButtonActivityIndicator` is now implemented inside the `ShippingLabelFormStepTableViewCell`, for showing an activity indicator on top of the `Continue` button when we are validating an address.
- `ShippingLabelSuggestedAddressTopBannerFactory`: the class that will return the top banner view inside the Shipping Label Suggested Address screen.
- Started to implement the `ShippingLabelSuggestedAddressViewController`. This is still a WIP and will be completed in the next PR.
- `ShippingLabelAddressFormViewController`: handled all the navigation things and removed the `topBanner` row which was unused. Implemented also the validation address for the origin and destination address, with the related navigation flow.
- `ShippingLabelFormViewModel`: declared a new struct called `State` which will handle all the necessary state to produce the ViewModel's outputs, including the sections. Implemented also the state machine flow to handle all the changes when row data is validated. Implemented the endpoint for validating an address directly in the main shipping label form.
- Added new tests.

## Testing
1. Launch the app in debug mode (because of the feature flag).
2. Open an order detail.
3. Press the button "Create Shipping Label".
4. Press the button "Continue" in the Create Shipping Label form. -> The activity indicator on top of the Continue button should appear.
5. If the address is valid, you should see the destination address with the **Continue** button enabled (in the origin cell the Continue button should disappear). If the address is not valid, you will navigate to the editing screen. If the address is valid but has a suggested address, you will navigate to the new screen with the suggested address (WIP).

## Screenshots
| Continue button pressed in Shipping Label Form           |  Shipping Label Suggested Address screen (WIP) |
:-------------------------:|:-------------------------:
![Simulator Screen Shot - iPhone 12 Pro - 2021-03-03 at 15 53 45](https://user-images.githubusercontent.com/495617/109824043-c16b9c80-7c38-11eb-9b43-84319e117130.png) | ![Simulator Screen Shot - iPhone 12 Pro - 2021-03-03 at 15 53 48](https://user-images.githubusercontent.com/495617/109824061-c3cdf680-7c38-11eb-984a-e6b2d6c49af0.png)


Update release notes:
